docs/tracing-context-propagation.md

```diff
@@ -0,0 +1,138 @@
+import cats.mtl.Local
+import org.typelevel.otel4s.java.context.Context
+import org.typelevel.otel4s.java.OtelJava
+import org.typelevel.otel4s.java.instances._ // brings Local derived from IOLocal
+import io.opentelemetry.api.GlobalOpenTelemetry
+
+def createOtel4s[F[_]: Async](implicit L: Local[F, Context]): Resource[F, OtelJava[F]] =
+  Resource
+    .eval(Async[F].delay(GlobalOpenTelemetry.get))
+    .map(OtelJava.local[F])
+    
+def program[F[_] : Async](otel4s: OtelJava[F]): F[Unit] = {
+  val _ = otel4s
+  Async[F].unit
+}
+
+val run: IO[Unit] =
+  IOLocal(Context.root).flatMap { implicit ioLocal: IOLocal[Context] =>
+    createOtel4s[IO].use(otel4s => program(otel4s))
+  }
+```
+
+If you don't need direct access to the `IOLocal` instance, there is also a shortcut `OtelJava.forAsync`:
+
+```scala mdoc:silent:reset
+import cats.effect._
+import org.typelevel.otel4s.java.OtelJava
+import io.opentelemetry.api.GlobalOpenTelemetry
+
+def createOtel4s[F[_]: Async: LiftIO]: Resource[F, OtelJava[F]] =
+  Resource
+    .eval(Async[F].delay(GlobalOpenTelemetry.get))
+    .evalMap(OtelJava.forAsync[F])
+
+def program[F[_]: Async](otel4s: OtelJava[F]): F[Unit] = {
+  val _ = otel4s
+  Async[F].unit
+}
+
+val run: IO[Unit] =
+  createOtel4s[IO].use(otel4s => program(otel4s))
+```
+
+### 2. Kleisli
+
+```scala mdoc:silent:reset
+import cats.effect._
+import cats.data.Kleisli
+import cats.mtl.Local
+import org.typelevel.otel4s.java.context.Context
+import org.typelevel.otel4s.java.OtelJava
+import io.opentelemetry.api.GlobalOpenTelemetry
+
+def createOtel4s[F[_]: Async](implicit L: Local[F, Context]): Resource[F, OtelJava[F]] =
+  Resource
+    .eval(Async[F].delay(GlobalOpenTelemetry.get))
+    .map(OtelJava.local[F])
+    
+def program[F[_]: Async](otel4s: OtelJava[F]): F[Unit] = {
+  val _ = otel4s
+  Async[F].unit
+}
+
+val kleisli: Kleisli[IO, Context, Unit] = 
+  createOtel4s[Kleisli[IO, Context, *]].use(otel4s => program(otel4s))
+
+val run: IO[Unit] = kleisli.run(Context.root)
+```
+
+## Limitations 
+
+The current encoding of [cats.effect.Resource][resource] is incompatible with `Local` semantics.
+For example, it's impossible to trace different stages of the resource (e.g. acquire, use, release) **and** 
+stay within the resource. A similar situation is with `fs2.Stream`.
+
+For example, you cannot get the following structure of the traces without allocation of the resource:
+```
+> resource
+  > acquire
+  > use
+    > inner spans
+  > release
+```
+
+To make it partially work, you need to trace each stage explicitly:
+```scala
+def acquire: F[Connection[F]] = ???
+def release(c: Connection[F]): F[Unit] = ???
+
+val resource: Resource[F, Connection[F]] = 
+  Resource.make(Tracer[F].span("acquire-connection").surround(acquire)) { connection =>
+    Tracer[F].span("release").surround(release(connection))
+  }
+
+val io: F[Unit] = Tracer[F].span("resource").surround(
+  resource.use(connection => Tracer[F].span("use").surround(fa(connection))) 
```

Comment:
`fa` does not appear to be defined anywhere

Reply:
Good catch, thanks!